src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -67,6 +67,37 @@ int64_t getAxisInRange(int64_t axis, Value val, bool includeRank) {
 // ONNX Op Shape Helper
 //===----------------------------------------------------------------------===//
 
+static void refineDimParams(
+    Operation *op, DimsExpr &inferredDims, Value output) {
+  // Get the index of output
+  if (!llvm::isa<OpResult>(output)) {
+    // output is a block parameter. It could be Func, Loop, If and etc.
+    // Give up now due to the complicated control flow.
+    return;
+  }
+
+  auto opResult = llvm::cast<OpResult>(output);
+  unsigned resultIndex = opResult.getResultNumber();
+  std::string dimParamsStr("");
+  bool isFirst = true;
+  for (unsigned i = 0; i < inferredDims.size(); ++i) {
+    if (inferredDims[i].isQuestionmark() && inferredDims[i].hasDimParam()) {
+      if (isFirst) {
+        isFirst = false;
+      } else {
+        dimParamsStr.append(",");
+      }
+      dimParamsStr = dimParamsStr + std::to_string(i) + ":" +
+                     inferredDims[i].getDimParam();
+    }
+  }
+  if (dimParamsStr == "")
+    return;
+  StringAttr dimParamsAttr = StringAttr::get(op->getContext(), dimParamsStr);
+  op->setAttr("onnx.dim_params_" + std::to_string(resultIndex),
```

Comment:
Suggestion: how about using `onnx.out_dim_params_`? I though `onnx.dim_params_0` is for the first input, but actually it is for the first output.